03-rnaseq/pathway-analysis_rnaseq_02_gsea.Rmd

```diff
@@ -392,11 +392,148 @@ head(lfc_vector)
 
 ### Run GSEA using the `GSEA()` function
 
-_Will be in upcoming PR_
+Genes were ranked from most highly positive to most highly negative, weighted according to their gene-level statistic, in the previous section.
+In this section, we will implement GSEA to calculate the enrichment score (ES) for each gene set using our pre-ranked gene list.
+
+We can use the `GSEA()` function to perform GSEA with any generic set of gene sets, but there are several functions for using specific, commonly used gene sets (e.g., `gseKEGG()`).
+
+```{r}
+gsea_results <- GSEA(
+  geneList = lfc_vector, # ordered ranked gene list
+  minGSSize = 25, # minimum gene set size
+  maxGSSize = 500, # maximum gene set set
+  pvalueCutoff = 0.05, # p value cutoff
+  eps = 0, # boundary for calculating the p value
+  seed = TRUE, # set seed to make results reproducible
+  pAdjustMethod = "BH", # Benjamini-Hochberg correction
+  TERM2GENE = dplyr::select(
+    mm_hallmark_sets,
+    gs_name,
+    gene_symbol
+  )
+)
+```
+
+The warning message above tells us that there are few genes that have the same log2 fold change value and they are therefore ranked equally.
+This means that `GSEA()` will arbitrarily choose which comes first in the ranked list [@GSEA-user-guide].
+This percentage of `0.19` is small, so we are not concerned that it will significantly effect our results.
+If the percentage was large on the other hand, we would be concerned about the log2 fold change results.
+
+Let's take a look at the results.
+
+```{r}
+# We can access the results from our `gsea_results` object using `@result`
+head(gsea_results@result)
+```
+
+Significance is assessed by permutating the gene labels of the pre-ranked gene list and recomputing the ES of the gene set for the permutated data, which generates a null distribution for the ES.
+The ES for each gene set is then normalized to account for the size of the set, resulting in a normalized enrichment score (NES), and an FDR (false discovery rate) value is calculated to account for multiple hypothesis testing.
+
+Looks like we have gene sets returned as significant at FDR of `0.05`, according to the `p.adjust` column in the table above.
+
+The information we're most likely interested in is in the `results` slot.
+Let's convert this into a data frame that we can use for further analysis and write to file.
+
+```{r}
+gsea_result_df <- data.frame(gsea_results@result)
+```
+
+## Visualizing results
+
+We can visualize GSEA results for individual pathways or gene sets using `enrichplot::gseaplot()`.
+Let's take a look at 2 different pathways -- one with a highly positive NES and one with a highly negative NES -- to get more insight into how ES are calculated.
+Note that if we didn't have any significant pathways, our visualizations below would show up blank as nothing would have met our `pvalueCutoff` above.
```

Comment:
I think this is only true for the dot plots? You are able to pull out pathways from `gsea_results` that don't meet the `pvalueCutoff` using `enrichplot::gseaplot()`.